sbndaq-artdaq/Generators/SBND/NevisTPC/NevisTB_generatorBase.cc

```diff
@@ -252,16 +257,53 @@ bool sbndaq::NevisTB_generatorBase::FillNTBFragment(artdaq::FragmentPtrs &frags,
   }  
   // Sweet, now, let's actually fill stuff                                                                                                              
   _this_event = ntbmetadata_.EventNumber();                                                                                                             
-                                                                                                                                                        
-  // set the subrun event 0 if it has never been set before                                                                                             
-  if (_subrun_event_0 == -1) {                                                                                                                          
-    _subrun_event_0 = _this_event;  
+          
+  //TLOG(TLVL_DEBUG+13) << "NTB Event number from pseudo counter " << _this_event;                                                                                                                                              
 
+  // set the subrun event 0 if it has never been set before
+  if (_subrun_event_0 == -1) {
+    _subrun_event_0 = _this_event; 
   }
     long long tframe   = ntbheader->getFrame();                  //Get trigger time tagged by nevis clock
     long long tsamp    = ntbheader->get2MHzSampleNumber();       //
     long long tdiv     = ntbheader->get16MHzRemainderNumber();   //
 
+    struct timespec unixtime;
+    clock_gettime(CLOCK_REALTIME, &unixtime);
+
+    //When the nevis clock rolls over (frames go back to 0) in between a PPS and a trigger
+    //catch the rollover by calculating the difference between PPS tagged time and trigger tagged time
+    int32_t   frame_diff = abs(ntbheader->getFrame() - GPSframe);
+    //long long tolerance  = 16777216 - FramesPerSecond_ - 10; //16777216 = 2^24 is when the rollover happens. frame numbers are stored in 24 bits
+    int32_t   tolerance  = 16770000;
+    int32_t   corrFrame;
+
+    TLOG(TLVL_INFO) << "Frame difference (between GPS and trigger): " << frame_diff;
+
+    if(tframe < prevFrame){
+      //If rollover happens, current frame number will be smaller than previous frame number. in that case, increment the rollover counter
+      rollCounter+=1;
+      TLOG(TLVL_INFO) << "Trigger frames rolled over!!!! " << " this many times: " << rollCounter;
+    }
+
+    corrFrame = tframe + rollCounter*16777216;
+    TLOG(TLVL_INFO) << " Corrected Frame:  " << corrFrame <<  " Uncorrected Frame: " << tframe;
```

Comment:
Can we change this to Debug? it spams the log messages pretty heavily

Reply:
would I still be able to see it in the log files?

Reply:
It should, but another option would be to make it a TRACE instead which would show up in logs but not the message viewer

Reply:
done